x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/RuleQueryBuilder.java

```diff
@@ -245,16 +240,35 @@ public void onFailure(Exception e) {
         });
 
         QueryBuilder newOrganicQuery = organicQuery.rewrite(queryRewriteContext);
-        RuleQueryBuilder rewritten = new RuleQueryBuilder(
-            newOrganicQuery,
-            matchCriteria,
-            this.rulesetId,
-            null,
-            null,
-            pinnedIdsSetOnce::get,
-            pinnedDocsSetOnce::get
-        );
+        List<String> idsToPin = pinnedIdsSetOnce.get();
+        List<Item> docsToPin = pinnedDocsSetOnce.get();
```

Comment:
`pinnedIdsSetOnce` will be set in the async call so we cannot access it here. We need to check the content in another rewrite round.

Reply:
Good catch, I've cleaned this up